### PR TITLE
Added TestContext.Write Is Obsolete Analyzer

### LINKED ISSUE #770: Add rule to detect calls to `TestContext.Write` methods and CodeFix to replace usages with `Out.Write`
We want to remove the `Write` method on `TestContext` which are just wrappers around the `Out` property.
To do this we need an analyzer rule with a code fix to detect usages and replace them with `TestContext.Out.Write`

See https://github.com/nunit/nunit/pull/4762

### PR DESCRIPTION
**Pull request by manfred-brands**
Fixes #770

### CODE REVIEW

File: documentation/NUnit1033.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+The `Write` methods are simple wrappers calling `Out.Write`.
+There is no wrapper for `Error` which always required to use `TestContext.Error.Write`.
+Besides this being inconsistent, later versions of .NET added new overloads,
+ e.g. for `ReadOnlySpan<char>` and `async` methods like `WriteAsync`.
+Instead of adding more and more dummy wrappers, it was decided that user code should use
+ the `Out` property and then can use any `Write` overload available on `TextWriter`.
+
+## How to fix violations
+
+Simple insert `.Out` between `TestContext` and `.Write`.
```
Comment (mikkelbu) [inline documentation/NUnit1033.md:28]:
```suggestion
Simply insert `.Out` between `TestContext` and `.Write`.
```

File: src/nunit.analyzers.tests/TestContextWriteIsObsolete/TestContextWriteIsObsoleteAnalyzerTests.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        private static readonly ExpectedDiagnostic expectedDiagnostic =
+            ExpectedDiagnostic.Create(AnalyzerIdentifiers.TestContextWriteIsObsolete);
+
+        [TestCaseSource(typeof(TestContextWriteIsObsoleteTestCases), nameof(TestContextWriteIsObsoleteTestCases.WriteInvocations))]
+        public void AnyDirectWriteMethod(string writeMethodAndParameters)
+        {
+            var testCode = TestUtility.WrapMethodInClassNamespaceAndAddUsings($@"
+                public void Test()
+                {{
+                    TestContext.{writeMethodAndParameters};
```
Comment (mikkelbu) [inline src/nunit.analyzers.tests/TestContextWriteIsObsolete/TestContextWriteIsObsoleteAnalyzerTests.cs:21]: Should we have a `↓` here? or will this have to be inside the `writeMethodAndParameters`

File: src/nunit.analyzers.tests/TestContextWriteIsObsolete/TestContextWriteIsObsoleteCodeFixTests.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        private static readonly ExpectedDiagnostic expectedDiagnostic =
+            ExpectedDiagnostic.Create(AnalyzerIdentifiers.TestContextWriteIsObsolete);
+
+        [TestCaseSource(typeof(TestContextWriteIsObsoleteTestCases), nameof(TestContextWriteIsObsoleteTestCases.WriteInvocations))]
+        public void AnyWriteMethod(string writeMethodAndParameters)
+        {
+            var code = TestUtility.WrapMethodInClassNamespaceAndAddUsings($@"
+                public void Test()
+                {{
+                    TestContext.{writeMethodAndParameters};
```
Comment (mikkelbu) [inline src/nunit.analyzers.tests/TestContextWriteIsObsolete/TestContextWriteIsObsoleteCodeFixTests.cs:23]: Same comment about `↓`

**Review by mikkelbu — APPROVED** [review]
LGTM. Only some minor nit-picks.

I've not had time to follow the discussion in the NUnit repo, but as far as I can tell these methods have not been marked as obsolete yet in the framework, so this Analyzer might confuse users by breaking the compilation - severity error - although the framework is fine, so perhaps this should just be warning (or perhaps we should rephrase the text a little to make it explicit, that these methods will be obsolete and then removed in future versions of the framework?)

**Comment by mikkelbu** [comment]
Just noticed that we have forgotten to update the sln file with the new markdown file

File: src/nunit.analyzers.tests/TestContextWriteIsObsolete/TestContextWriteIsObsoleteAnalyzerTests.cs (hunk 2 above)
Comment (manfred-brands) [reply src/nunit.analyzers.tests/TestContextWriteIsObsolete/TestContextWriteIsObsoleteAnalyzerTests.cs]: Done.

File: src/nunit.analyzers.tests/TestContextWriteIsObsolete/TestContextWriteIsObsoleteCodeFixTests.cs (hunk 3 above)
Comment (manfred-brands) [reply src/nunit.analyzers.tests/TestContextWriteIsObsolete/TestContextWriteIsObsoleteCodeFixTests.cs]: Done.

File: documentation/NUnit1033.md (hunk 1 above)
Comment (manfred-brands) [reply documentation/NUnit1033.md]: Done.

**Comment by manfred-brands** [comment]
> I've not had time to follow the discussion in the NUnit repo, but as far as I can tell these methods have not been marked as obsolete yet in the framework, so this Analyzer might confuse users by breaking the compilation - severity error - although the framework is fine, so perhaps this should just be warning (or perhaps we should rephrase the text a little to make it explicit, that these methods will be obsolete and then removed in future versions of the framework?)

I have both change the level to `Warning` and rephrased it from `are Obsolete` into `will be marked as Obsolete and eventually removed`

**Pull request merged**